### DISCUSSION TITLE
How to import a JSON file?

### DISCUSSION THREAD

**Discussion by gryzzly** [preactjs/wmr#474] [Q&A]
With `wmr 1.3.2`:

I’m trying to import a JSON file in which the contents for the website are stored. 

I place the `content.json` in `public` folder next to my `index.js` file. In the `index.js` file I have
```
import content from './content.json';
```

This ^ works in development mode, the file’s imported and content is available. 

In build mode, it fails like this:
`"FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory"`
<details>
<summary>Click to see full error</summary>

```
> @ build /Users/misha/Source/website
> wmr build --prerender


<--- Last few GCs --->

[56697:0x1046c2000]    42853 ms: Scavenge (reduce) 2046.4 (2050.3) -> 2045.8 (2051.5) MB, 6.5 / 0.0 ms  (average mu = 0.209, current mu = 0.163) allocation failure 
[56697:0x1046c2000]    45887 ms: Mark-sweep (reduce) 2046.6 (2050.5) -> 2046.0 (2052.0) MB, 3033.2 / 0.0 ms  (+ 0.1 ms in 127 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 3095 ms) (average mu = 0.138, current mu = 0.0

<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x1012e4da5 node::Abort() (.cold.1) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
 2: 0x1000a6239 node::Abort() [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
 3: 0x1000a639f node::OnFatalError(char const*, char const*) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
 4: 0x1001e9057 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
 5: 0x1001e8ff3 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
 6: 0x100397ee5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
 7: 0x1003999aa v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
 8: 0x100395079 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
 9: 0x100392911 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
10: 0x1003a11aa v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
11: 0x1003a1231 v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
12: 0x10036ebd7 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
13: 0x1006ed918 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
14: 0x100a7a279 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
15: 0x100ab474a Builtins_PerformPromiseThen [/Users/misha/.nvm/versions/node/v14.15.4/bin/node]
Abort trap: 6
```
</details>

If I use the [undocumented?] `json:` prefix(not sure where I could find it, only saw reading through other issues that it’s a possibility) I get the same error.

However, when using `\0json:` prefix, like this:
```
import content from '\0json:./content.json';
```

I’m getting a different result: 
`'json:./content.json' is imported by public/index.js, but could not be resolved – treating it as an external dependency`
<details>
  <summary>Click to see full error</summary>

```
'json:./content.json' is imported by public/index.js, but could not be resolved – treating it as an external dependency

Wrote 39Kb to disk:
  index.html 469b
  admin/login.html 165b
  admin/index.html 16Kb
  index.dac70bac.js 16Kb
  chunks/prerender.e3e1cd05.js 4.9Kb
  assets/style.e6ee4bcf.css 466b
Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only file and data URLs are supported by the default ESM loader. Received protocol 'json:'
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:782:11)
    at Loader.resolve (internal/modules/esm/loader.js:86:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:230:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:56:40)
    at link (internal/modules/esm/module_job.js:55:36)
```
</details>

**Comment by gryzzly** [comment]
I mentioned this here, but I’m not sure if it’s directly related 
https://github.com/preactjs/wmr/issues/449#issuecomment-808751404

**Reply by marvinhagemeister** [reply]
Importing `*.json` files should just work out of the box. So this sounds like a bug to me.

**Reply by danielweck** [reply]
I too can reproduce this JSON bug.
I tried both bare `*.json` import, as well as the `json:` import prefix.
`npm start` development mode works fine, but `npm run build --prerender` production mode goes into an infinite loop of doom.
My current workaround is to use an ad-hoc `text:` plugin, to load JSON as well as SVG.
See:
https://github.com/preactjs/wmr/issues/449#issuecomment-808465192
Important: I am forced to use different file extensions (`*.json` => `*.jsonld` and `*.svg` => `*.xml`) in order to work around another related bug in WMR's plugin system.

Note that I stopped using the official WMR release and I now build my app directly with WMR's `main` branch (so I can experiment with the latest "bleeding edge" updates).

I see that there is an upcoming PR with further changes to the plugin system, so I look forward to testing this :)

https://github.com/preactjs/wmr/pull/479

**Reply by gryzzly** [reply]
> Importing `*.json` files should just work out of the box. So this sounds like a bug to me.

Would you like me to report this as an issue?

**Reply by gryzzly** [reply]
> My current workaround is to use an ad-hoc `text:` plugin, to load JSON as well as SVG.
> See:
> [#449 (comment)](https://github.com/preactjs/wmr/issues/449#issuecomment-808465192)
> Important: I am forced to use different file extensions (`*.json` => `*.jsonld` and `*.svg` => `*.xml`) in order to work around another related bug in WMR's plugin system.

Thanks for the suggestion! I’ve tried using `text:` without success, but it didn’t occur to me to try using a different file extension. Feels a bit too much of a workaround though, if a fix is planned I’d rather wait for it.

**Reply by danielweck** [reply]
> Would you like me to report this as an issue?

I forgot to mention that I had already filed this one :)
https://github.com/preactjs/wmr/issues/446

Closed now. However, I am getting regression bugs in other features. Investigating ...

**Reply by danielweck** [reply]
> Closed now. However, I am getting regression bugs in other features. Investigating ...

Apologies for raising a false alarm, I had a HTTP server / cache issue.
No regressions.
(the `text:` / `other:` -prefixed imports still fail when referencing WMR-handled file extensions such as SVG and JSON, but that's a different issue, see https://github.com/preactjs/wmr/issues/449 )
